<a href="https://colab.research.google.com/github/sigvehaug/Introduction-to-Python-for-Medical-Researchers/blob/master/08-Numpy-images.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Introduction to Python Programming for Medical Researchers, University of Bern, Sigve Haug

This notebook is a slightly revised version of G. Witz' course [Image Processing with Python](https://github.com/guiwitz/Python_image_processing).

# 8. Numpy with (Medical) Images

The Numpy documentation is [here](https://numpy.org/doc/stable/index.html#numpy-docs-mainpage).

---

All images are essentially matrices with a variable number of dimensions where each element represents the value of one pixel. The different dimensions and the pixel values can have very different meanings depending on the type of image considered, but the structure is the same.

Python does not allow by default to gracefully handle multi-dimensional data. In particular it is not desgined to handle matrix operations. Numpy was developed to fill in this blank and offers a very similar framework as the one offered by Matlab. It is underlying a large number of packages and has become abolsutely essential to Python scientific programming. In particular it underlies the functions of scikit-image. The latter in turn forms the basis of other software like CellProfiler. It is thus essential to have a good understanding of Numpy to  proceed.

Instead of introducing Numpy in an abstract way, we are going here to present it through the lense of image processing in order to focus on the most useful features in the context of this course.

## 8.1 Exploring an image

Some test images are provided directly in skimage, so let us look at one (we'll deal with the details of image import later). First let us import the necessary packages.

In [ ]:
# This would a standard example image in the skimage package
#image = skimage.data.coins()

We retrieve a skin lesion image from the github repository of the course.

In [ ]:
url = 'https://raw.githubusercontent.com/sigvehaug/Introduction-to-Python-for-Medical-Researchers/master/Data/ISIC_0000015.jpg'


In [ ]:
# What is the shape of the image ?


(767, 1022, 3)

We see that this image has 767 rows, 1022 columns and 3 (RGB) channels. Let us now work on only one channel.

The first thing we can do is check the type of the image:

We thus see that the imported image is indeed a Numpy array. We can also find out what is the type of each pixel:

We see here that the image is an 8bit image. Classical format we are going to see are 8bit (uint8), 16bit (uint16) and non-integers (usually float64). The type of the image pixels set what values they can take. For example 8bit means values from $0$ to $2^8 -1= 256-1 = 255$.

Since typing is dynamic in Python, if we for example divide our uint8 image by 2 it ends up being a float.

In [ ]:
image2.dtype

dtype('float64')

Let us check what are the image dimensions using the shape method:

We see that the image has 767 rows and 1022 columns. Let us have a look at it:

Let us also look at the actual matrix behind that image:

Jupyter only shows us a part of this large matrix, adding ... between the first and last few rows/columns. Each value here corresponds to a single pixel of the image. The type of data contained in the array, here 8bit integer, is also displayed.

## 8.2 Operations on arrays

### 8.2.2 Arithmetic operations

Two types of simple arithmetic operations can be done on arrays. One can modify them using a scalar, for example by adding a constant:

One has to be very careful with the type of the image.

As the array has 8bit type, with vlaues ranging from 0 to 255, any value above 255 limit gets reassigned at the start of the range.

It is important to note that when doing an operation, Python adopts the most complex type of the variables involved (called upcasting). Hence if we add a float instead of an integer:

the result is now correct and is of type float. If needed, arrays can be converted to approriate formats using the astype() method:

[[485. 469. 485. ... 484. 470. 484.]
 [464. 272. 239. ... 307. 324. 465.]
 [485. 230. 230. ... 298. 306. 484.]
 ...
 [485. 255. 234. ... 314. 314. 484.]
 [459. 267. 230. ... 323. 338. 474.]
 [485. 470. 485. ... 474. 460. 484.]]


The second type of operation is element-wise by using two arrays of the same dimension. Let's change the image type to avoid the type of issue mentionded above.

Now we can for example multiply the image by itself:

### 8.2.3 Logical operations

A set of important operations when processing images are logical (or boolean) operations. Those have a very simple syntax in Numpy. For example, let's compare pixel intensities to some value *a*:

(767, 1022)

dtype('bool')

We see that the returned object is a logical matrix (containg True and False) of the same size as the image. That returned object can be directly assigned to a variable, leading to a very compact formulation:

Of course other logical operator can be used (<, >, ==, !=) and the resulting boolean matrices combined:

## 8.2.4 Functional operations

Numpy has a large trove of functions to act on arrays. Just like the arithmetic operations described above, those functions are designed to be applied to entire arrays, and can either generate a new array or a single value. For example the mean, maximum and minmal pixel values can be abtained like this:

Alternatively one can take the square root of each array element by doing:

It is impossible to give an exhaustive presentation of Numpy functions. We will simply describe functions that we use a we go through course material.

## 8.3 Slicing and indexing

Let us now go back to our color image to explore how to select specific parts of an image.

We see that the image has three dimensions, probably it's a stack of three images of size 767x1022. Let us try to have a look at this image hoping that dimensions are handled gracefully:

The image being in natural colors, the three dimensions probably indicate an RGB (red, green, blue) format, and the plotting function just knows what to do in that case.

### 8.3.1 Array slicing

Let us now just look at one of the three planes composing the image. To do that, we are going the select a portion of the image array by slicing it. One can give:
- a single index e.g. 0 for the first element
- a range e.g. 0:10 for the first 10 elements
- take all elements using a semi-column :

What portion is selected has to be specified for each dimensions of an array. In our particular case, we want to select all rows, all columns and a single plane of the image:

We see now the red layer of the image. We can do the same for the others by specifying planes 0, 1, and 2:

Logically intensities are different. We can confirm that by measuring the mean of each plane. To do that we use the same function as above but apply it to a singel sliced plane:

In [ ]:
image4 = image[:,:,0]

In [ ]:
np.mean(image4)

174.28380071286966

and for all planes using a comprehension list:

To look at some more details let us focus on a smaller portion of the image e.g. one of the lesion core. For that we are going to take a slice of the red image and store it in a new variable and display the selection. We consider pixel rows from 200 to 600 and columns from 300 to 700 of the first plane (0).

There are different ways to select parts of an array. For example one can select every n'th element by giving a step size. In the case of an image, this subsamples the data:

Or use negative indices or steps to start from the end of the array. In our case this flips dimensions:

### 8.3.2 Array indexing

In addition to slicing an array, we can also select specific values out of it. There are [many](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html) different ways to achieve that, but we focus here on two main ones.

First one might have a list of pixel positions and one wishes to get the values of those pixels. By passing two lists of the same size containing the rows and columns positions of those pixels, one can recover them:

Alternatively, one can pass a logical array of the same dimensions as the original array, and only the True pixels are selected. For example, let us create a logical array by picking values above a threshold:

Let's visualize it. Matplotlib handles logical arrays simply as a binary image:

We can recover the value of all the "white" (True) pixels in the original image by using:

And now ask how many pixels are above threshold and what their average value is.

We now know that there are 2585 pixels above the threshold and that their mean is 153.6

## 8.4 Creating arrays

Often we are going to create new arrays that we are going to then transform. There are a few common ways of doing that.

Creating a zero array of dimensions 3x5:

Same for an array filled with ones:

Create an array filled with random uniform (0-1) values:

Clearly we did something wrong. The argument that was passed was not accepted by the functions. This is a good occasion to look at the help for a more complex function, in order to know how to pass arguments.

So we understand now that we can just pass a list of numbers:

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

## 8.5 Assembling arrays

It might happen that we have separate images that need to be assembled into one larger matrix. Numpy provides several functions to achieve that. For example we can concatenate images, which means that we append an image to another. Let's grab each channel of the image:

And concateante them:

We can specify along which axis they should be assembled:

In [ ]:
concatenated_image = np.concatenate((image_red,image_green),axis = 1)

Alternatively, we might want to create a stack of image from separate images. Let's recreate the cat image by switching the red and blue channel. We assemble the separate color images and specific that they should be stacked along the third axis (axis =2).

In [ ]:
switch_image = np.stack((image_blue,image_green,image_red),axis = 2)

We verify that the dimensions are the expected ones: